File: noor-shield-pc/src/main/license.js
'use strict';

const crypto = require('crypto');
const { SUPABASE_URL, SUPABASE_ANON_KEY } = require('../../service/supabaseConfig');

/**
 * Product-key activation, checked against a real Supabase table
 * (license_keys / activate_license_key — see cloud/schema.sql) so a key
 * already claimed by another PC is actually rejected, not just a made-up
 * or mistyped one. Deliberately "check once, then work offline forever":
 * activateKeyOnline() is only ever called at the moment the parent clicks
 * "Activate" — once store.activated is true, the filter keeps enforcing
 * with no further need for internet, matching the rest of the app.
 *
 * Keys are hashed before ever leaving this PC — only the SHA-256 hash is
 * sent, never the plaintext key.
 */

const TRIAL_DAYS = 7;
const TRIAL_MS = TRIAL_DAYS * 24 * 60 * 60 * 1000;

/**
 * Days left in the one-time free trial, counted from the first time the
 * protection service ever ran on this PC (store.firstRunAt) — not from
 * install, so reinstalling the app alone can't be used to keep resetting
 * it. 0 once the trial has run out. Always 0 if firstRunAt isn't set yet
 * (the caller is expected to set it before relying on this).
 */
function trialDaysRemaining(firstRunAt, now = Date.now()) {
  if (!firstRunAt) return 0;
  const remainingMs = TRIAL_MS - (now - firstRunAt);
  return Math.max(0, Math.ceil(remainingMs / (24 * 60 * 60 * 1000)));
}

function isTrialActive(firstRunAt, now = Date.now()) {
  return Boolean(firstRunAt) && now - firstRunAt < TRIAL_MS;
}

/**
 * True if an activated key is still within its access window. `expiresAt`
 * is null for a lifetime key (the common case) — always active once
 * activated. A time-limited key (e.g. a 29-day key) carries a real
 * timestamp, set once at activation time (see activate_license_key in
 * schema.sql) and stored locally in store.license.expiresAt, checked here
 * with no further network round trip — same offline-first design as the
 * rest of activation, just no longer "forever" for these keys.
 */
function isLicenseActive(activated, expiresAt, now = Date.now()) {
  if (!activated) return false;
  if (!expiresAt) return true;
  return now < new Date(expiresAt).getTime();
}

/** "NOOR-XXXXX-XXXXX-XXXXX" -> "NOORXXXXXXXXXXXXXXX", tolerant of case/whitespace/missing dashes. */
function normalizeKey(rawInput) {
  return String(rawInput || '')
    .toUpperCase()
    .replace(/[^A-Z0-9]/g, '');
}

/** Hashes a key the same way scripts/generate-keys.js did, or null if the format is wrong. */
function hashKey(rawInput) {
  const normalized = normalizeKey(rawInput);
  if (normalized.length !== 19 || !normalized.startsWith('NOOR')) return null;
  const body = normalized.slice(4);
  const canonical = `NOOR-${body.slice(0, 5)}-${body.slice(5, 10)}-${body.slice(10, 15)}`;
  return crypto.createHash('sha256').update(canonical).digest('hex');
}

/**
 * Checks a key against Supabase's activate_license_key function, scoped to
 * this PC's deviceId (see store.js's `license.deviceId`, generated once and
 * reused on every later activation attempt — including reinstalls — so the
 * same PC re-activating its own key never looks like reuse on another PC).
 *
 * Returns { ok: true, expiresAt } on success — expiresAt is null for a
 * lifetime key or an ISO timestamp for a time-limited one (see
 * isLicenseActive) — or { ok: false, reason } where reason is 'invalid'
 * (not a real key), 'already_used' (claimed by a different PC),
 * 'bad_format' (didn't even look like a key), or 'network_error' (couldn't
 * reach Supabase at all).
 */
async function activateKeyOnline(rawInput, deviceId) {
  const hash = hashKey(rawInput);
  if (!hash) return { ok: false, reason: 'bad_format' };

  let res;
  try {
    res = await fetch(`${SUPABASE_URL}/rest/v1/rpc/activate_license_key`, {
      method: 'POST',
      headers: {
        apikey: SUPABASE_ANON_KEY,
        Authorization: `Bearer ${SUPABASE_ANON_KEY}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ p_key_hash: hash, p_device_id: deviceId }),
    });
  } catch {
    return { ok: false, reason: 'network_error' };
  }

  if (!res.ok) return { ok: false, reason: 'network_error' };

  const result = JSON.parse(await res.text());
  if (result.status === 'ok') return { ok: true, expiresAt: result.expires_at };
  if (result.status === 'already_used') return { ok: false, reason: 'already_used' };
  return { ok: false, reason: 'invalid' };
}

module.exports = {
  normalizeKey,
  hashKey,
  activateKeyOnline,
  trialDaysRemaining,
  isTrialActive,
  isLicenseActive,
  TRIAL_DAYS,
};

File: noor-shield-app/app/src/main/java/com/barnoota/noorshield/license/License.kt
package com.barnoota.noorshield.license

import com.barnoota.noorshield.cloud.CloudConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.security.MessageDigest
import java.util.concurrent.TimeUnit

/**
 * Kotlin port of noor-shield-pc/src/main/license.js: product-key activation
 * checked against the same Supabase table (license_keys /
 * activate_license_key) so a key already claimed by another device is
 * actually rejected. "Check once, then work offline forever" — same as the
 * PC app: activateKeyOnline() only runs when the parent taps "Activate";
 * once activated, the filter keeps enforcing with no further internet need.
 */
object License {
    const val TRIAL_DAYS = 29
    private const val TRIAL_MS = TRIAL_DAYS * 24L * 60 * 60 * 1000

    private val http = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .build()
    private val JSON = "application/json; charset=utf-8".toMediaType()

    fun trialDaysRemaining(firstRunAt: Long?, nowMs: Long = System.currentTimeMillis()): Int {
        if (firstRunAt == null) return 0
        val remainingMs = TRIAL_MS - (nowMs - firstRunAt)
        return maxOf(0, Math.ceil(remainingMs / (24.0 * 60 * 60 * 1000)).toInt())
    }

    fun isTrialActive(firstRunAt: Long?, nowMs: Long = System.currentTimeMillis()): Boolean =
        firstRunAt != null && nowMs - firstRunAt < TRIAL_MS

    /**
     * True if an activated key is still within its access window. `expiresAtMs`
     * is null for a lifetime key — always active once activated. A
     * time-limited key (e.g. a 29-day key) carries a real timestamp, set
     * once at activation time (see activate_license_key in schema.sql) and
     * checked here with no further network round trip.
     */
    fun isLicenseActive(activated: Boolean, expiresAtMs: Long?, nowMs: Long = System.currentTimeMillis()): Boolean {
        if (!activated) return false
        if (expiresAtMs == null) return true
        return nowMs < expiresAtMs
    }

    /** "NOOR-XXXXX-XXXXX-XXXXX" -> "NOORXXXXXXXXXXXXXXX", tolerant of case/whitespace/missing dashes. */
    fun normalizeKey(rawInput: String): String = rawInput.uppercase().filter { it.isLetterOrDigit() }

    /** Hashes a key the same way scripts/generate-keys.js did, or null if the format is wrong. */
    fun hashKey(rawInput: String): String? {
        val normalized = normalizeKey(rawInput)
        if (normalized.length != 19 || !normalized.startsWith("NOOR")) return null
        val body = normalized.substring(4)
        val canonical = "NOOR-${body.substring(0, 5)}-${body.substring(5, 10)}-${body.substring(10, 15)}"
        val digest = MessageDigest.getInstance("SHA-256").digest(canonical.toByteArray(Charsets.UTF_8))
        return digest.joinToString("") { "%02x".format(it) }
    }

    sealed class ActivationResult {
        /** expiresAtMs is null for a lifetime key, or the key's expiry for a time-limited one (e.g. a 29-day key). */
        data class Ok(val expiresAtMs: Long?) : ActivationResult()
        data object BadFormat : ActivationResult()
        data object Invalid : ActivationResult()
        data object AlreadyUsed : ActivationResult()
        data object NetworkError : ActivationResult()
    }

    /**
     * Checks a key against Supabase's activate_license_key function, scoped
     * to this device's license deviceId (CloudStore.licenseDeviceId — reused
     * on every later activation attempt, including reinstalls, so the same
     * device re-activating its own key never looks like reuse on another
     * device).
     */
    suspend fun activateKeyOnline(rawInput: String, deviceId: String): ActivationResult = withContext(Dispatchers.IO) {
        val hash = hashKey(rawInput) ?: return@withContext ActivationResult.BadFormat

        val body = JSONObject().put("p_key_hash", hash).put("p_device_id", deviceId)
        val request = Request.Builder()
            .url("${CloudConfig.URL}/rest/v1/rpc/activate_license_key")
            .addHeader("apikey", CloudConfig.ANON_KEY)
            .addHeader("Authorization", "Bearer ${CloudConfig.ANON_KEY}")
            .post(body.toString().toRequestBody(JSON))
            .build()

        try {
            http.newCall(request).execute().use { response ->
                if (!response.isSuccessful) return@withContext ActivationResult.NetworkError
                val text = response.body?.string().orEmpty()
                val json = JSONObject(text)
                when (json.optString("status")) {
                    "ok" -> {
                        val expiresAt = json.optString("expires_at", null)
                        val expiresAtMs = expiresAt?.let { runCatching { java.time.Instant.parse(it).toEpochMilli() }.getOrNull() }
                        ActivationResult.Ok(expiresAtMs)
                    }
                    "already_used" -> ActivationResult.AlreadyUsed
                    else -> ActivationResult.Invalid
                }
            }
        } catch (e: Exception) {
            ActivationResult.NetworkError
        }
    }
}
